File: js/interfaces/default/views/packing.js
// Trip Packing Lists — a checklist per trip. "Auto-suggested" is a set of
// built-in templates (weekend / beach / ski / international) that bulk-add
// common items in one click; nothing here calls out to any external service.

import { el, fmtDate } from '../dom.js';

let state = { selectedId: null };

const TEMPLATES = {
  weekend: {
    label: 'Weekend trip',
    items: [
      ['Clothing', ['Underwear', 'Socks', 'Sleepwear', 'Casual outfit']],
      ['Toiletries', ['Toothbrush', 'Toothpaste', 'Deodorant']],
      ['Documents', ['ID / license', 'Wallet']],
      ['Electronics', ['Phone charger']],
    ],
  },
  beach: {
    label: 'Beach / warm',
    items: [
      ['Clothing', ['Swimsuit', 'Sandals', 'Sunhat', 'Light clothing']],
      ['Toiletries', ['Sunscreen', 'After-sun lotion', 'Toothbrush']],
      ['Gear', ['Beach towel', 'Sunglasses', 'Reusable water bottle']],
      ['Documents', ['ID / license']],
    ],
  },
  ski: {
    label: 'Ski / cold',
    items: [
      ['Clothing', ['Thermal base layers', 'Winter jacket', 'Gloves', 'Wool socks', 'Beanie']],
      ['Gear', ['Goggles', 'Hand warmers']],
      ['Toiletries', ['Lip balm', 'Moisturizer', 'Toothbrush']],
      ['Documents', ['ID / license']],
    ],
  },
  international: {
    label: 'International',
    items: [
      ['Documents', ['Passport', 'Visa (if needed)', 'Travel insurance', 'Copies of documents']],
      ['Electronics', ['Power adapter', 'Phone charger', 'Offline maps downloaded']],
      ['Clothing', ['Underwear', 'Socks', 'Versatile outfits']],
      ['Toiletries', ['Toothbrush', 'Toothpaste (travel size)']],
      ['Money', ['Local currency', 'Backup card']],
    ],
  },
};

function itemRow(item, ctx, rerender) {
  return el('div', { class: 'mer-task-row' }, [
    el('input', {
      type: 'checkbox', checked: item.packed,
      onclick: async (e) => { await ctx.data.PackingItems.update(item.id, { packed: e.target.checked }); rerender(); },
    }),
    el('span', { class: item.packed ? 'mer-task-title is-done' : 'mer-task-title', text: item.name || '(untitled)' }),
    el('button', {
      type: 'button', class: 'mer-icon-btn', text: '×', title: 'Remove',
      onclick: async () => { await ctx.data.PackingItems.remove(item.id); rerender(); },
    }),
  ]);
}

async function renderListDetail(canvas, list, ctx, rerender) {
  canvas.append(el('div', { class: 'mer-detail-header' }, [
    el('h1', { text: list.name || '(untitled trip)' }),
    el('button', { type: 'button', class: 'mer-icon-btn', text: '✕ Close', onclick: () => { state.selectedId = null; rerender(); } }),
  ]));
  if (list.tripDate) canvas.append(el('p', { class: 'mer-muted', text: fmtDate(list.tripDate) }));

  canvas.append(el('div', { class: 'mer-subsection-label', text: 'Add from a template' }));
  const templateRow = el('div', { class: 'mer-toolbar' }, Object.entries(TEMPLATES).map(([key, tpl]) =>
    el('button', {
      type: 'button', text: tpl.label,
      onclick: async () => {
        for (const [category, names] of tpl.items) {
          for (const name of names) {
            await ctx.data.PackingItems.create({ listId: list.id, name, category, packed: false });
          }
        }
        rerender();
      },
    })));
  canvas.append(templateRow);

  const nameIn = el('input', { type: 'text', placeholder: 'Add an item…' });
  const catIn = el('input', { type: 'text', placeholder: 'Category (optional)' });
  canvas.append(el('div', { class: 'mer-subsection-label', text: 'Add an item' }));
  canvas.append(el('div', { class: 'mer-person-form' }, [nameIn, catIn, el('button', {
    type: 'button', text: 'Add',
    onclick: async () => {
      if (!nameIn.value.trim()) return;
      await ctx.data.PackingItems.create({ listId: list.id, name: nameIn.value.trim(), category: catIn.value.trim() || 'Other', packed: false });
      nameIn.value = ''; catIn.value = '';
      rerender();
    },
  })]));

  const items = await ctx.data.PackingItems.byIndex('listId', list.id);
  const packedCount = items.filter((i) => i.packed).length;
  canvas.append(el('div', { class: 'mer-subsection-label', text: `Packed ${packedCount} / ${items.length}` }));

  if (!items.length) {
    canvas.append(el('p', { class: 'mer-muted', text: 'Nothing on this list yet — add items or use a template above.' }));
    return;
  }

  const byCategory = new Map();
  for (const item of items) {
    const cat = item.category || 'Other';
    if (!byCategory.has(cat)) byCategory.set(cat, []);
    byCategory.get(cat).push(item);
  }
  for (const [cat, catItems] of byCategory) {
    canvas.append(el('div', { class: 'mer-subsection-label', text: cat }));
    const wrap = el('div', { class: 'mer-task-list-area' });
    for (const item of catItems) wrap.append(itemRow(item, ctx, rerender));
    canvas.append(wrap);
  }
}

function listCard(list, progress, onSelect, ctx, rerender) {
  const card = el('div', { class: 'mer-place-card' }, [
    el('div', { class: 'mer-place-body' }, [
      el('div', { class: 'mer-place-name', text: list.name || '(untitled trip)' }),
      el('div', { class: 'mer-place-meta' }, [
        list.tripDate ? el('span', { class: 'mer-chip', text: fmtDate(list.tripDate) }) : null,
        el('span', { class: 'mer-chip', text: `${progress.packed}/${progress.total} packed` }),
      ].filter(Boolean)),
    ]),
    el('button', {
      type: 'button', class: 'mer-icon-btn', text: '×', title: 'Delete list',
      onclick: async (e) => {
        e.stopPropagation();
        if (!confirm(`Delete "${list.name}" and all its items?`)) return;
        const items = await ctx.data.PackingItems.byIndex('listId', list.id);
        for (const item of items) await ctx.data.PackingItems.remove(item.id);
        await ctx.data.PackingLists.remove(list.id);
        rerender();
      },
    }),
  ]);
  card.addEventListener('click', () => onSelect(list.id));
  return card;
}

export async function renderPacking(canvas, ctx, rerender) {
  if (state.selectedId) {
    const list = await ctx.data.PackingLists.get(state.selectedId);
    if (list) return renderListDetail(canvas, list, ctx, rerender);
    state.selectedId = null;
  }

  canvas.append(el('h1', { text: 'Trip Packing Lists' }));
  canvas.append(el('p', { class: 'mer-muted', text: 'One checklist per trip, with templates to get started fast.' }));

  const nameIn = el('input', { type: 'text', placeholder: 'Trip name (e.g. "Tokyo")' });
  const dateIn = el('input', { type: 'date' });
  canvas.append(el('div', { class: 'mer-subsection-label', text: 'New trip' }));
  canvas.append(el('div', { class: 'mer-person-form' }, [nameIn, dateIn, el('button', {
    type: 'button', text: 'Create list',
    onclick: async () => {
      if (!nameIn.value.trim()) return;
      const list = await ctx.data.PackingLists.create({ name: nameIn.value.trim(), tripDate: dateIn.value || null });
      state.selectedId = list.id;
      rerender();
    },
  })]));

  const lists = await ctx.data.PackingLists.list();
  canvas.append(el('div', { class: 'mer-subsection-label', text: `Your trips (${lists.length})` }));
  if (!lists.length) {
    canvas.append(el('p', { class: 'mer-muted', text: 'No trips yet.' }));
    return;
  }
  const itemLists = await Promise.all(lists.map((l) => ctx.data.PackingItems.byIndex('listId', l.id)));
  const grid = el('div', { class: 'mer-place-grid' });
  lists.forEach((l, i) => {
    const items = itemLists[i];
    const progress = { packed: items.filter((it) => it.packed).length, total: items.length };
    grid.append(listCard(l, progress, (id) => { state.selectedId = id; rerender(); }, ctx, rerender));
  });
  canvas.append(grid);
}

File: native/composeApp/src/commonMain/kotlin/com/alekpeed/lifeos/packing/Packing.kt
package com.alekpeed.lifeos.packing

import com.alekpeed.lifeos.Storage
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

// Packing lists — one checklist per trip, items grouped by category, with templates
// that bulk-add common items. Persists as one JSON blob under "Packing"; old flat stubs
// migrate into a single list.
//
// Absorbed into Travel (§5.1): a packing list is meaningless outside a trip, so it is
// reached through the trip rather than from its own nav slot. The data deliberately did
// NOT move — lists stay under the "Packing" key and gained an optional `tripId` instead.
// Migrating them into TravelData would have rewritten every existing list to gain a
// nav change, and a list belonging to no trip would have had nowhere to land. Old JSON
// still loads (the field defaults), and a list with tripId 0 shows as unassigned.

@Serializable
data class PackItem(val id: Long, val name: String, val category: String = "Other", val packed: Boolean = false)

@Serializable
data class PackingList(
    val id: Long,
    val name: String,
    val tripDate: String = "",
    val items: List<PackItem> = emptyList(),
    // 0 = not attached to a trip.
    val tripId: Long = 0L,
)

// A saved template. Same shape as the built-ins, but stored, so "save this list as a
// template" is one button (§5.1).
@Serializable
data class PackTemplateGroup(val category: String, val items: List<String> = emptyList())

@Serializable
data class PackTemplate(val id: Long, val name: String, val groups: List<PackTemplateGroup> = emptyList())

@Serializable
data class PackingData(
    val lists: List<PackingList> = emptyList(),
    val templates: List<PackTemplate> = emptyList(),
)

// The built-in starting points, always offered alongside any you save yourself.
//
// name -> list of (category, itemNames). Nothing calls out anywhere; pure data.
val PACKING_TEMPLATES: List<Pair<String, List<Pair<String, List<String>>>>> = listOf(
    "Weekend trip" to listOf(
        "Clothing" to listOf("Underwear", "Socks", "Sleepwear", "Casual outfit"),
        "Toiletries" to listOf("Toothbrush", "Toothpaste", "Deodorant"),
        "Documents" to listOf("ID / license", "Wallet"),
        "Electronics" to listOf("Phone charger"),
    ),
    "Beach / warm" to listOf(
        "Clothing" to listOf("Swimsuit", "Sandals", "Sunhat", "Light clothing"),
        "Toiletries" to listOf("Sunscreen", "After-sun lotion", "Toothbrush"),
        "Gear" to listOf("Beach towel", "Sunglasses", "Reusable water bottle"),
        "Documents" to listOf("ID / license"),
    ),
    "Ski / cold" to listOf(
        "Clothing" to listOf("Thermal base layers", "Winter jacket", "Gloves", "Wool socks", "Beanie"),
        "Gear" to listOf("Goggles", "Hand warmers"),
        "Toiletries" to listOf("Lip balm", "Moisturizer", "Toothbrush"),
        "Documents" to listOf("ID / license"),
    ),
    "International" to listOf(
        "Documents" to listOf("Passport", "Visa (if needed)", "Travel insurance", "Copies of documents"),
        "Electronics" to listOf("Power adapter", "Phone charger", "Offline maps downloaded"),
        "Clothing" to listOf("Underwear", "Socks", "Versatile outfits"),
        "Toiletries" to listOf("Toothbrush", "Toothpaste (travel size)"),
        "Money" to listOf("Local currency", "Backup card"),
    ),
)

private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

fun loadPacking(): PackingData {
    val raw = Storage.read("Packing")
    if (raw.isNullOrBlank()) return PackingData()
    if (raw.trimStart().startsWith("{")) {
        return runCatching { json.decodeFromString<PackingData>(raw) }.getOrElse { PackingData() }
    }
    // Old SimpleListScreen stub: flat item names — fold into one list.
    val names = raw.lines().filter { it.isNotBlank() }
    if (names.isEmpty()) return PackingData()
    val items = names.mapIndexed { i, n -> PackItem(id = i + 2L, name = n.trim()) }
    return PackingData(lists = listOf(PackingList(id = 1L, name = "My packing list", items = items)))
}

fun savePacking(data: PackingData) {
    Storage.write("Packing", json.encodeToString(data))
}

fun packingListsFor(tripId: Long): List<PackingList> =
    loadPacking().lists.filter { it.tripId == tripId }

// Lists belonging to no trip. They stay reachable after the module lost its nav slot.
fun unassignedPackingLists(): List<PackingList> = loadPacking().lists.filter { it.tripId == 0L }

fun PackingList.packedCount(): Int = items.count { it.packed }

// Built-ins and saved templates in one list, so a caller never has to know which is which.
fun allTemplates(): List<PackTemplate> {
    val builtIn = PACKING_TEMPLATES.mapIndexed { i, (name, groups) ->
        PackTemplate(
            id = -(i + 1L), // negative ids: built-ins are not stored and cannot be deleted
            name = name,
            groups = groups.map { (cat, items) -> PackTemplateGroup(cat, items) },
        )
    }
    return builtIn + loadPacking().templates
}

// Turn a list into a reusable template. Packed state is deliberately dropped — a
// template is what to bring, not what you already put in the bag.
fun saveAsTemplate(list: PackingList, name: String) {
    val data = loadPacking()
    val groups = list.items.groupBy { it.category }
        .map { (cat, items) -> PackTemplateGroup(cat, items.map { it.name }) }
    val id = (data.templates.maxOfOrNull { it.id } ?: 0L) + 1
    savePacking(data.copy(templates = data.templates + PackTemplate(id, name.trim().ifBlank { list.name }, groups)))
}

fun deleteTemplate(id: Long) {
    val data = loadPacking()
    savePacking(data.copy(templates = data.templates.filterNot { it.id == id }))
}
